aria/orchestrator/execution_plugin/ctx_proxy/server.py

```diff
@@ -150,43 +150,101 @@ def __exit__(self, *args, **kwargs):
+        modifying_key = None
+        modifying_value = None
+
+    num_args = len(args)
+
     while index < num_args:
         arg = args[index]
+
+        # Object attribute
         attr = _desugar_attr(current, arg)
-        if attr:
+        if attr is not None:
             current = getattr(current, attr)
-        elif isinstance(current, collections.MutableMapping):
-            key = arg
-            path_dict = _PathDictAccess(current)
-            if index + 1 == num_args:
-                # read dict prop by path
-                value = path_dict.get(key)
-                current = value
-            elif index + 2 == num_args:
-                # set dict prop by path
-                value = args[index + 1]
-                path_dict.set(key, value)
-                current = None
-            else:
-                raise RuntimeError('Illegal argument while accessing dict')
-            break
+
+        # List entry
+        elif isinstance(current, list) and _is_int(arg):
+            current = current[int(arg)]
+
+        # Dict (or dict-like object) value
+        elif hasattr(current, '__getitem__'):
+            if modifying and (not arg in current):
+                current[arg] = {}
+            current = current[arg]
+
+        # Call
         elif callable(current):
-            kwargs = {}
-            remaining_args = args[index:]
-            if isinstance(remaining_args[-1], collections.MutableMapping):
-                kwargs = remaining_args[-1]
-                remaining_args = remaining_args[:-1]
-            current = current(*remaining_args, **kwargs)
-            break
+            if isinstance(current, functools.partial):
+                argspec = inspect.getargspec(current.func)
+                # Don't count initial args fulfilled by the partial
+                spec_args = argspec.args[len(current.args):]
+                # Don't count keyword args fulfilled by the partial
+                spec_args = tuple(a for a in spec_args if a not in current.keywords)
+            else:
+                argspec = inspect.getargspec(current)
+                spec_args = argspec.args
+
+            callable_kwargs = {}
+            if isinstance(arg, dict):
+                # If the first arg is a dict, treat it as our kwargs
+                # TODO: what if the first argument really needs to be a dict?
+                callable_kwargs = arg
+                index += 1
+
+            if argspec.varargs is not None:
+                # Gobble the rest of the args
+                callable_args = args[index:]
+            else:
+                # Take only what we need
+                spec_args = tuple(a for a in spec_args if a not in callable_kwargs)
+                spec_args_count = len(spec_args)
+                if inspect.ismethod(current):
+                    # Don't count "self" argument
+                    spec_args_count -= 1
+                callable_args = args[index:index + spec_args_count]
```

Comment:
See comment above: we can't just grab all remaining args, because we might need these args for further delving into the returned value of the callable.